### ISSUE TITLE
Error when I'm trying to dispatch an action outside of my React Component, the state isn't updating

### ISSUE THREAD

**Issue by DevSebas-Code** [reduxjs/redux-toolkit#2563]
Hi, I'm trying to dispatch an action from outside React's logic.

This  app is running with Electron.

**Context of the problem**:  In this app I'm receiving push notifications through a third-party-service (Pushy), so we need to dispatch an action which set the new state en our reducer , so we can handle the info of the notification in a React Component later.

The data is being received in the main electron file (main.ts)

// _Imports of my main.ts_
```
import { app, BrowserWindow,  screen } from 'electron';
import installExtension, {
  REACT_DEVELOPER_TOOLS,
} from 'electron-devtools-installer';
import Pushy from 'pushy-electron';
import { setData } from './redux/reducers/setDataReducer';
import {store} from './redux/store';
```

This is how i define my dispatch method:
```
const dispatch = store.dispatch
```

This is the structure of the data that I'm trying to save in the store: 

```
export interface DataReceived {
  id: string;
  message: [];
}
```

_This is the portion of code where I'm trying to dispatch my action:_

** Pushy.setNotificationListener()  is a Listener function** 

Inside the scope of this function I'm  trying to dispatch my action setData(), declared in my reducer.
```
      Pushy.setNotificationListener((data: DataReceived) => {
        
        const auxData = JSON.stringify(data);
  
        
        dispatch(setData(data))

        // Display an alert with the "message" payload value
        Pushy.alert(mainWindow, 'Received notification: ' + auxData);
      });

```
**My reducer**: 

**setDataReducer.ts**

Once my action is dispatched from my main.ts file I can see that the payload has the data that i need to set in my new state of the reducer

```
import { createSlice } from '@reduxjs/toolkit';

interface DataReceived {
  id: string;
  message: [];
}

let initialState: DataReceived = {
  id: null,
  message: [],
};

export const dataSetter = createSlice({
  name: 'data',
  initialState,
  reducers: {
    setData(state, action) {
      console.log("action.payload: ", action.payload)

      // Result of console.log()
      // *********************
      // action.payload:  {
      //   id: 'd1b42186-64dc-42ca-b88f-64e40c58a1ef',
      //   message: '[{"story":{"id":"f53505c7-ae3c-4390-875a-e65b434b28f5","name":"Soy una historia","title":"Soy un titulo","description":"Soy una descripción","image":"https://storage.googleapis.com/vixonic-dashbX|oard/center","createdAt":2131221313},"tags":[]}]'
      // 

      return {
        ...state,
        id: action.payload.id,
        message: action.payload.message,
      };
    },
  },
});

export const { setData } = dataSetter.actions;

export default dataSetter.reducer;

```

**store.ts** 
```

import  {configureStore}  from "@reduxjs/toolkit";
import { useDispatch } from "react-redux";
import { dataSetter } from "./reducers/setDataReducer";

export const store = configureStore({
    reducer: dataSetter.reducer
})

export type RootState = ReturnType<typeof store.getState>

export type AppDispatch  = typeof store.dispatch

export const useAppDispatch : () => AppDispatch = useDispatch

export const selectStory = (state: RootState) => state


```
As the documentation suggests, i declared my hooks in a separate file to avoid circular import 

**hooks.ts**
```
 
import {useDispatch, useSelector } from 'react-redux'
import type { TypedUseSelectorHook } from 'react-redux'
import type { RootState, AppDispatch } from './store'

// Use throughout your app instead of plain `useDispatch` and `useSelector`
export const useAppDispatch: () => AppDispatch = useDispatch
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector


```

This is my React component,  the const story, is my selector of the state, and I'm passing as an argument the function selectStory() imported from the store file, this function basically is selecting the current state of the Redux store.

I'm trying to render the selected state in a **p** html tag, **and only returns me the initial state of the reducer, and not the new one which is the one that I'm receiving in the payload**

```
import React from 'react';
import { Grid } from '@mui/material';
import StoriesCarousel from '@src/components/Stories/StoriesCarousel';
import { useAppSelector, useAppDispatch } from '../redux/hooks'
import { selectStory } from '@src/redux/store'; 


export interface IStoriesProps {}



const Stories: React.FunctionComponent<IStoriesProps> = (props) => {

  const story = useAppSelector(selectStory)
  const stories: any[] = [
    {
      content: () => (
        
        <div style={{ display: 'flex', flex: 1, justifyContent: 'center' }}>
          <p>{JSON.stringify(story)}</p>
        </div>
      ),
    },
  ];
  return (
    <Grid container item display='flex' justifyContent='center'>
      <StoriesCarousel stories={stories} />
    </Grid>
  );
};

export default Stories;
```
This is the graphic representation of my problem 
 
![image](https://user-images.githubusercontent.com/82928758/182548164-a44b921d-c718-4567-9e80-389f48063fc2.png)

I'll be very glad if someone could help me, or telling me if the dispatch method is correctly used.

**Regards**

**Comment by markerikson** [comment]
I"m afraid it's kind of hard to follow what's going on without seeing an actual repo+codebase.

First question: is the action actually reaching the reducer when you dispatch? Does that `console.log()` run? Do the Redux DevTools show the action in the history list?

Assuming that the action actually _does_ show up:

I know that Electron splits things into multiple processes. Is it possible that you're somehow creating _two_ different Redux stores and using them in different parts of the app?

**Comment by DevSebas-Code** [comment]
Hi @markerikson, thanks for your reply, the console.log() of the reducer is running, it shows me the payload with the data that I'm trying to set.

This is the result of the console.log() : 

      // action.payload:  {
      //   id: 'd1b42186-64dc-42ca-b88f-64e40c58a1ef',
      //   message: '[{"story":{"id":"f53505c7-ae3c-4390-875a-e65b434b28f5","name":"Soy una historia","title":"Soy un titulo","description":"Soy una descripción","image":"https://storage.googleapis.com/vixonic-dashbX|oard/center","createdAt":2131221313},"tags":[]}]'
      //

And for the second question that you mentioned, i don't think that could be viable, considering that one of the main principles of Redux is the single source of truth, ref : https://redux.js.org/understanding/thinking-in-redux/three-principles.

Questions : Should allow Redux Toolkit dispatch actions from outside of a component?
Do you think that I'm selecting the data correctly?

**Comment by markerikson** [comment]
@DevSebas-Code I'm not talking about "Redux principles" here.

I'm saying that it's _possible_ that, due to how Electron is architected, you _might_ be creating _two_ different store instances without even intending to.  _If_ that's what's happening, then it's also possible that you might be dispatching an action to store A, but using store B in the components, and thus they would never seen an update.

Redux absolutely allows dispatching outside of a component. Anywhere you have access to `dispatch()`, you can use it. You said that the action is showing up in the reducer, so we _know_ that the dispatch step is working okay.  I can see that the reducer is doing an immutable update (which btw you don't have to write by hand with RTK, and I have a couple other nitpicks with the code, but that's not the issue atm).

So, there has to be something wrong with the overall Electron+Redux setup instead.

**Comment by DevSebas-Code** [comment]
About your first paragraph: I understand, you're saying that because of the Electron multi processes the store that I'm calling from my front-end is only showing the initial state, because the state really mutates in the store instance of the Electron process.

Due this, Could it be a good option to declare the entire store config in my electron main file?

And if not, Is there any specific configuration for the store in Electron apps?

In relation with the inmutable update:  (Only if you cloud asnwer me, if not there's no problem) ; 

What would be the best practice to avoid write by hand that? (Any docs ref?)

Again, thanks for your support until now, means a lot ! :smile:

**Comment by markerikson** [comment]
Unfortunately I've never used Electron myself, so I don't have a good answer here.

I do know there's tutorials and boilerplates out there that show how to add Redux to an Electron app - the only real suggestion that I can offer atm is to go read through some of those and see what's different from your current setup.

**Comment by DevSebas-Code** [comment]
Hi @markerikson, thanks for your support, i solved my problem, using the ipcMain and ipcRenderer functions for Electron, So then i could dispatch the action to the front end.

Main.ts (Electron file)
```

 Pushy.setNotificationListener((data: DataReceived) => {
      const { ipcMain } = require('electron');
      console.log('Data Sended: ', data);
      ipcMain.on('variable-request', function (event, arg) {
        console.log('SENDING....', arg);
        event.sender.send('variable-reply', [data]);
      });
```

Renderer.tsx (Where my react app is rendered)

```
import React from 'react';
import ReactDOM from 'react-dom';
import Application from './components/Application';
import { inDev } from './utils/helpers';
import Wrapper from './components/Wrapper';
import { store } from './redux/store';
import { Provider } from 'react-redux';
import { setData } from './redux/reducers/SetVixonicData';

const dispatch = store.dispatch;
const ipcRendererExec = async () => {
  setInterval(() => {
    const ipcRenderer = window.require('electron').ipcRenderer;
    ipcRenderer.send('variable-request', ['somevar', 'anothervar']);
    ipcRenderer.on('variable-reply', function (event, args) {
      console.log('event: ', event);
      dispatch(setData(args));
      console.log('args: ', args);
    });
    console.log('After Function oooo');
  }, 60000);
};

ipcRendererExec();

// Say something
console.log('[ERWT] : Renderer execution started');

// Application to Render
const app = (
  <Provider store={store}>
    <Wrapper>
      <Application title='hola' version='1.0' />
    </Wrapper>
  </Provider>
);

// // store.dispatch({})

// Hot module replacement
if (inDev() && module.hot) module.hot.accept();

// window.Electron.ipcRenderer.once(CHANNEL)


// Render application in DOM
ReactDOM.render(app, document.getElementById('app'));
```
